File: parte2_plataforma/comun/esquema.py
"""Esquema canónico de un viaje y reglas de validación (versión Python).

Las reglas están en `config/esquema_viaje.json`, que también lee el código Scala de Spark
(`pids.Esquema`). Así el histórico y el tiempo real aplican exactamente las mismas reglas, como
exige E3, y los tests de los dos lenguajes comprueban los mismos resultados sobre la muestra.

Los datos llegan con nombres y formatos distintos según la fuente:
  - exportación CSV de NYC Open Data: VendorID, fechas "01/01/2020 12:28:15 AM"
  - API de NYC Open Data (SODA):      vendorid, fechas ISO, todo como texto
  - Parquet mensual de la TLC:         VendorID, fechas ya como timestamp, alguna columna extra
`normalizar` lo lleva todo a los nombres canónicos y a tipos numéricos.

Semántica de nulos (igual en Scala): un campo vacío solo invalida el viaje si es obligatorio
(regla `falta_campo_obligatorio`); el resto de reglas no se consideran incumplidas por un nulo.
"""
from __future__ import annotations

import json
import os
from functools import lru_cache
from pathlib import Path

import pandas as pd

RUTA_CONFIG = Path(os.environ.get('PIDS_CONFIG_DIR', Path(__file__).resolve().parents[2] / 'config'))


@lru_cache(maxsize=1)
def config() -> dict:
    return json.loads((RUTA_CONFIG / 'esquema_viaje.json').read_text(encoding='utf-8'))


def columnas() -> list[str]:
    return list(config()['alias'].values())


def normalizar(df: pd.DataFrame) -> tuple[pd.DataFrame, list[str]]:
    """Renombra a los nombres canónicos y convierte tipos.

    Devuelve el DataFrame canónico y las columnas de la fuente que no se reconocen (se ignoran,
    pero conviene saberlo: puede ser un cambio de formato del proveedor). Los valores que no se
    pueden convertir quedan como nulos y los detecta `validar`.
    """
    cfg = config()
    renombrar, extra = {}, []
    for col in df.columns:
        canon = cfg['alias'].get(str(col).strip().lower())
        if canon is None:
            extra.append(str(col))
        else:
            renombrar[col] = canon
    out = df.rename(columns=renombrar)[list(renombrar.values())].copy()
    for col in columnas():
        if col not in out.columns:
            out[col] = pd.NA
    for col in cfg['fechas']:
        out[col] = _a_fecha(out[col])
    for col in columnas():
        if col in cfg['fechas'] or col in cfg['texto']:
            continue
        valores = _a_numero(out[col], cfg.get('decimal_coma', False))
        out[col] = valores.astype('Int64') if col in cfg['enteras'] and _son_enteros(valores) else valores
    for col in cfg['texto']:
        out[col] = out[col].astype('string').str.strip().str.upper()
    return out[columnas()], extra


def _son_enteros(valores: pd.Series) -> bool:
    v = valores.dropna()
    return bool((v == v.round()).all())


def _a_numero(serie: pd.Series, decimal_coma: bool) -> pd.Series:
    """A número, aceptando el decimal con coma de la exportación completa ("1,2" o "1.234,56")."""
    if pd.api.types.is_numeric_dtype(serie):
        return serie
    texto = serie.astype('string').str.strip()
    if decimal_coma:
        con_coma = texto.str.contains(',', na=False)
        if con_coma.any():
            # con coma decimal, el punto es separador de miles
            texto = texto.mask(con_coma, texto.str.replace('.', '', regex=False).str.replace(',', '.', regex=False))
    return pd.to_numeric(texto, errors='coerce')


def _a_fecha(serie: pd.Series) -> pd.Series:
    """A datetime sin zona horaria y con la misma resolución, probando los formatos de cada fuente.

    Muestra de Moodle "01/01/2020 12:28:15 AM", exportación completa "2020 Jan 01 12:28:15 AM",
    API en ISO y Parquet ya como timestamp.
    """
    if pd.api.types.is_datetime64_any_dtype(serie):
        return (serie.dt.tz_localize(None) if serie.dt.tz is not None else serie).astype('datetime64[us]')
    texto = serie.astype('string').str.strip()
    fechas = pd.Series(pd.NaT, index=serie.index, dtype='datetime64[us]')
    for formato in config()['formatos_fecha_python'] + ['ISO8601']:
        pendientes = fechas.isna() & texto.notna()
        if not pendientes.any():
            break
        intento = pd.to_datetime(texto[pendientes], format=formato, errors='coerce')
        fechas[pendientes] = intento.astype('datetime64[us]')
    return fechas.astype('datetime64[us]')


def validar(df: pd.DataFrame) -> tuple[pd.DataFrame, pd.DataFrame]:
    """Separa los viajes válidos de los rechazados.

    Devuelve (validos, rechazados). `rechazados` lleva una columna `motivos` con la lista de
    reglas incumplidas, en el mismo orden que en Spark.
    """
    r = config()['reglas']
    duracion = df['llegada'] - df['recogida']
    zona_ok = range(r['zona_minima'], r['zona_maxima'] + 1)

    def fuera(col, validos):
        return df[col].notna() & ~df[col].isin(validos)

    reglas = {
        'falta_campo_obligatorio': df[config()['obligatorias']].isna().any(axis=1),
        'fecha_fuera_de_rango': df['recogida'].notna() & ~df['recogida'].between(
            pd.Timestamp(r['recogida_desde']), pd.Timestamp(r['recogida_hasta']), inclusive='left'),
        'duracion_no_positiva': duracion <= pd.Timedelta(0),
        'duracion_excesiva': duracion > pd.Timedelta(hours=r['duracion_maxima_horas']),
        'distancia_fuera_de_rango': df['distancia_millas'].notna()
                                    & ~df['distancia_millas'].between(0, r['distancia_maxima_millas']),
        'importe_negativo': (df['tarifa'] < 0) | (df['importe_total'] < 0),
        'vendor_desconocido': fuera('vendor_id', r['vendor_id']),
        'zona_desconocida': fuera('zona_origen', zona_ok) | fuera('zona_destino', zona_ok),
        'tipo_pago_desconocido': fuera('tipo_pago', r['tipo_pago']),
        'tarifa_desconocida': fuera('tarifa_id', r['tarifa_id']),
        'indicador_invalido': fuera('almacenado_y_reenviado', r['almacenado_y_reenviado']),
    }
    fallos = pd.DataFrame({k: v.astype('boolean').fillna(False).astype(bool) for k, v in reglas.items()},
                          index=df.index)
    malo = fallos.any(axis=1)
    rechazados = df[malo].copy()
    rechazados['motivos'] = [[k for k, v in fila.items() if v] for fila in fallos[malo].to_dict('records')]
    return df[~malo].copy(), rechazados


def avisos(df: pd.DataFrame) -> dict[str, int]:
    """Valores sospechosos que NO invalidan el viaje pero conviene vigilar."""
    return {
        'pasajeros_cero': int((df['pasajeros'] == 0).sum()),
        'pasajeros_nulos': int(df['pasajeros'].isna().sum()),
        'distancia_cero': int((df['distancia_millas'] == 0).sum()),
        'tarifa_cero': int((df['tarifa'] == 0).sum()),
    }


def nombre(campo: str, codigo) -> str:
    """Significado de un código según el diccionario de la TLC."""
    if pd.isna(codigo):
        return '?'
    return config()['diccionario'].get(campo, {}).get(str(int(codigo)), '?')

File: parte2_plataforma/spark/src/main/scala/pids/Esquema.scala
package pids

import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql.{Column, DataFrame}

/** Esquema canónico y validación de viajes (versión Spark de parte2_plataforma/comun/esquema.py).
  *
  * La misma función sirve para el histórico (CSV/Parquet) y para los eventos en tiempo real: es la
  * garantía de que los dos caminos aplican exactamente las mismas reglas (E3).
  *
  * Semántica de nulos (igual que en Python): un campo vacío solo invalida el viaje si es obligatorio.
  */
object Esquema {
  val Motivos = "motivos"

  /** Renombra al esquema canónico y convierte tipos. Los valores no convertibles quedan como nulos.
    * Los códigos numéricos se dejan como Double hasta validar (ver `tipar`).
    *
    * @param conservar columnas de la fuente que se mantienen tal cual (procedencia, por ejemplo)
    */
  def normalizar(df: DataFrame, cfg: ConfigEsquema, conservar: Seq[String] = Nil): DataFrame = {
    val porNombre = df.schema.fields.map(f => f.name.trim.toLowerCase -> f).toMap
    val columnas = cfg.alias.map { case (fuente, canon) =>
      porNombre.get(fuente) match {
        case None    => lit(null).cast(tipoDestino(canon, cfg)).as(canon)
        case Some(f) => convertir(f, canon, cfg).as(canon)
      }
    }
    df.select(columnas ++ conservar.map(c => col(s"`$c`")): _*)
  }

  private def tipoDestino(canon: String, cfg: ConfigEsquema): DataType =
    if (cfg.fechas(canon)) TimestampType else if (cfg.texto(canon)) StringType else DoubleType

  private def convertir(f: StructField, canon: String, cfg: ConfigEsquema): Column = {
    val c = col(s"`${f.name}`")
    val n = s"`${f.name}`"
    if (cfg.fechas(canon)) f.dataType match {
      case TimestampType | TimestampNTZType | DateType => c.cast(TimestampType)
      case _ =>
        // cada fuente trae su formato: muestra de Moodle, exportación completa (mes abreviado) o ISO
        val intentos = cfg.formatosFecha.map(fmt => s"try_to_timestamp(trim($n), '$fmt')") :+
          s"try_to_timestamp(trim($n))"
        expr(intentos.mkString("coalesce(", ", ", ")"))
    }
    else if (cfg.texto(canon)) upper(trim(c.cast(StringType)))
    else f.dataType match {
      case _: NumericType => c.cast(DoubleType)
      case _ if cfg.decimalComa =>
        // "1,2" y "1.234,56": con coma decimal, el punto es separador de miles
        expr(s"""case when trim($n) like '%,%'
                 |     then try_cast(replace(replace(trim($n), '.', ''), ',', '.') AS DOUBLE)
                 |     else try_cast(trim($n) AS DOUBLE) end""".stripMargin)
      case _ => expr(s"try_cast(trim($n) AS DOUBLE)")
    }
  }

  /** Añade la columna `motivos` con las reglas incumplidas (vacía si el viaje es válido). */
  def validar(df: DataFrame, cfg: ConfigEsquema): DataFrame = {
    val duracion = expr("unix_seconds(llegada) - unix_seconds(recogida)")
    def fueraDeLista[T](c: String, validos: Seq[T]): Column = not(col(c).isin(validos: _*))
    def zonaInvalida(c: String): Column =
      col(c) < cfg.zonaMinima || col(c) > cfg.zonaMaxima || col(c) =!= floor(col(c))

    val reglas: Seq[(String, Column)] = Seq(
      "falta_campo_obligatorio" -> cfg.obligatorias.map(c => col(c).isNull).reduce(_ || _),
      "fecha_fuera_de_rango" -> (col("recogida") < to_timestamp(lit(cfg.recogidaDesde)) ||
        col("recogida") >= to_timestamp(lit(cfg.recogidaHasta))),
      "duracion_no_positiva" -> (duracion <= 0),
      "duracion_excesiva" -> (duracion > cfg.duracionMaximaHoras * 3600),
      "distancia_fuera_de_rango" -> (col("distancia_millas") < 0 || col("distancia_millas") > cfg.distanciaMaxima),
      "importe_negativo" -> (col("tarifa") < 0 || col("importe_total") < 0),
      "vendor_desconocido" -> fueraDeLista("vendor_id", cfg.vendors),
      "zona_desconocida" -> (zonaInvalida("zona_origen") || zonaInvalida("zona_destino")),
      "tipo_pago_desconocido" -> fueraDeLista("tipo_pago", cfg.tiposPago),
      "tarifa_desconocida" -> fueraDeLista("tarifa_id", cfg.tarifas),
      "indicador_invalido" -> fueraDeLista("almacenado_y_reenviado", cfg.indicadores)
    )
    val motivos = array_compact(array(reglas.map { case (nombre, incumple) =>
      when(coalesce(incumple, lit(false)), lit(nombre))
    }: _*))
    df.withColumn(Motivos, motivos)
  }

  /** Tipos definitivos para guardar: los códigos pasan a entero (ya validados). */
  def tipar(df: DataFrame, cfg: ConfigEsquema): DataFrame =
    cfg.enteras.foldLeft(df)((d, c) => d.withColumn(c, col(c).cast(IntegerType)))

  /** Separa (válidos, rechazados) a partir de la columna `motivos`. */
  def separar(validado: DataFrame, cfg: ConfigEsquema): (DataFrame, DataFrame) = {
    val validos = tipar(validado.filter(size(col(Motivos)) === 0).drop(Motivos), cfg)
    val rechazados = validado.filter(size(col(Motivos)) > 0)
    (validos, rechazados)
  }
}
